Mission 2

codingjamee opened the pull request:
- UI로직과 domain로직을 나누어서 반영 
 - 시작은 index.js로 시작하게 하였습니다. 
   - UI로직은 사용자와의 인터렉션에 관련이 있는 코드를 작성하였습니다. 
     - 사용자 입력을 받는다
     - 최소 최대값 입력을 받는다 
     - 재시작 종료 옵션
     - 몇회만에 맞췄는지 보여준다
   - domain로직은 계산하고 결과를 보여주는 로직을 작성하였습니다. 
     - 계산하고 결과를 출력하는 로직
     - 최소 최대 값 사이의 랜덤숫자 생성
     - 업 다운 정답 계산하여 출력
     - 이전 추측들을 보여준다 

1. UI에서의 의도한 사항
  - 자주쓰이는 함수는 별도로 재사용하기 위해 생성하였습니다. 
  - 너무 길어지는 함수도 가독성 향상을 위해 별도의 함수로 나누었습니다. 
  - 중복되는 코드에서의 다른 코드 (contant의 상수값)는 별도의 객체로 분리하였고
    프로퍼티에 할당하여 쉽게 읽히도록 하였습니다. 
2. domain에서 의도한 사항 
  - 하나의 함수 안에 내포한 함수로 한 덩어리로 분류하고자 하였습니다.
  - UI로직에 의존성을 제거하기 위해 유효하지 않은 배열인 경우 push메서드 작동 배제하였습니다.

codingjamee committed:
mission1요구사항

codingjamee committed:
mission1 -> mission2 기능 추가 요구사항

codingjamee committed:
domain과 UI로직 분리 초안

codingjamee committed:
재사용성을 위해 startSetting을 함수로 분리
tryLoop도 같은 위계를 맞추기 위해 분리
자주 쓰이는 validtaion 숫자임을 확인하기 위한 함수 분리

codingjamee committed:
showInput로직 domain에 추가

codingjamee committed:
console두번 출력되는 것 수정

codingjamee committed:
정답 콘솔 주석

codingjamee committed:
의존성 분리를 위해
array일 때만 push메서드 사용

codingjamee committed:
에러 수정'

codingjamee committed:
min max설정,userInputSet함수 분리

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+/**
```
P5; ASK;

소소한 궁금증인데 UI는 대문자, domain은 소문자로 네이밍 하신 이유가 있을까요?

codingjamee replied:
UI는 줄임말이고 domain은 줄이지 않아서? 라고 생각했는데 지금 다시보니 조금 어색해보이긴 하네요.  
폴더는 보통 소문자로 많이 사용하는데 ui라고 쓰는게 더 덜 어색할까요?

eastroots92 replied:
아니에요!

말씀하신 것 처럼 UI와 같은 축약어는 대문자를 쓰는 경우가 많아요. (특히 변수명 지을때)
폴더 명에서 대문자는 살짝 어색하긴한데 지극히 주관적인 부분이라고 생각해요.

단지 의도를 가지고 한 것일까? 이 부분이 궁금해서 여쭈어보았던 거였어요!!
만약 UI라는 단어가 어색하다면 View, Page 와 같은 네이밍도 많이 사용되는 것 같아요!

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+계산 로직과 다른점은
+계산은 사용자와의 인터렉션에 관련이 없으나
+UI는 사용자와의 인터렉션에 관련이 있음
+그러므로 count에 따라 보여줘야할지 말지는 UI에서 다뤄야 한다고 판단했음
```
아주 굿입니다!! :+1:

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+*/
+
+export async function playGame() {
+  let { answer, trialLimit } = await startSetting();
```
P2;

여기서 const가 아닌 let을 사용한 이유가 있을까요? 
const로도 문제 없이 잘 동작 할 것 같아서요!

codingjamee replied:
오 넵 이부분 수정하겠습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+  trial = await tryLoop(trial, trialLimit, answer);
+}
+
+async function startSetting() {
+  const boundary = await userInputSet(constant.askBoundary);
+  const [minNumber, maxNumber] = setMinMax(boundary);
+  const trialLimit = await userInputSet(constant.askTrialLimit);
+
+  const answer = gameData(minNumber, maxNumber).answer;
+  console.log(
+    `컴퓨터가 ${minNumber}~${maxNumber} 사이의 숫자를 선택했습니다. 숫자를 맞춰보세요.`
+  );
+  return { answer, trialLimit };
+}
```
P3; ASK;

입력 출력을 처리하는 부분과, `gameData` 로직을 구분하는 접근법은 매우 좋다고 생각해요! :+1:

다만 `startSetting` 를 비롯한 `validation` 등등의 함수들을 보았을 때 값을 출력하는 것과 처리된 결과를 return 하는 것이 하나의 함수 안에 있는 것이 좋을까? 하는 생각이 들었어요.
관점에 따라 다르겠지만 어떻게 보면 두가지 동작이 하나의 함수에 있다고 보여져서요.

어색하다 생각한 점.
1. 함수를 사용하면서  둘다 동작하는 것을 기대할 수 있을까? 함수를 보았을 때 값을 return 받거나, 처리에 대한 출력이 있거나 둘 중 하나를 기대하고 호출 할 것 같은데 두개의 동작이 있는 것이 어색하다.
2. 만약 어떤 경우에는 값을 출력하고, 어떤 상황에서는 출력하지 않고 싶다면 이 함수에서는 어떻게 처리할 수 있을까? (이 함수 내에 컨텍스트를 마구마구 추가해야 할 것 같은데 🤔)

codingjamee replied:
한 함수에서 result와 isValid값을 담은 객체를 return하는 클로저를 작성해보았습니다. 
이렇게 되면 result와 isValid를 따로따로 활용할 수 있을 것 같아 나누어 받는 쪽에서 처리하도록 할 수 있을 것 같아서
이렇게 작성해보았습니다.

codingjamee replied:
> 어색하다 생각한 점.
> 
> 1. 함수를 사용하면서  둘다 동작하는 것을 기대할 수 있을까? 함수를 보았을 때 값을 return 받거나, 처리에 대한 출력이 있거나 둘 중 하나를 기대하고 호출 할 것 같은데 두개의 동작이 있는 것이 어색하다.
> 2. 만약 어떤 경우에는 값을 출력하고, 어떤 상황에서는 출력하지 않고 싶다면 이 함수에서는 어떻게 처리할 수 있을까? (이 함수 내에 컨텍스트를 마구마구 추가해야 할 것 같은데 🤔)

한 함수에서 result와 isValid값을 담은 객체를 return하는 클로저를 작성해보았습니다. 
이렇게 되면 result와 isValid를 따로따로 활용할 수 있을 것 같아 나누어 받는 쪽에서 처리하도록 할 수 있을 것 같아서
이렇게 작성해보았습니다.

eastroots92 commented on src/domain/index.js:
```diff
@@ -0,0 +1,46 @@
+    }
+  }
+
+  function showInput(guessArr, guessInput) {
+    const guess = guessArr;
+    const toNumberInput = parseFloat(guessInput);
+    if (Array.isArray(guess)) guess.push(toNumberInput);
+    console.log(toNumberInput);
+    console.log("이전 추측:", guess.join(", "));
+    return guess;
+  }
```
P1;

console.log는 UI와 관련된 부분으로 생각되는데 어떻게 생각하시나요?

제가 생각했을 때 Domain / UI를 나누는 기준이 UI가 console에서 web으로 바뀌었을 때.
정말 잘 설계가 되었다면 Domain의 변경은 없거나, 최소화 되어야 한다고 생각해요.

codingjamee replied:
> P1;
> 
> console.log는 UI와 관련된 부분으로 생각되는데 어떻게 생각하시나요?
> 
> 제가 생각했을 때 Domain / UI를 나누는 기준이 UI가 console에서 web으로 바뀌었을 때. 정말 잘 설계가 되었다면 Domain의 변경은 없거나, 최소화 되어야 한다고 생각해요.

네 이부분 반영하여 결괏값을 return하도록 로직 변경하여 UI쪽에서 받아서 처리하였습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+}
+
+async function userInputSet(askConstant) {
+  const result = await readLineAsync(askConstant);
+  const input = result.split(",");
```
ask;

result에서 split `,`을 하는 이유가 있을까요?

codingjamee replied:
만일 a,b형태로 user의 복수 개의 숫자를 받는다면 split을 해서 array로 사용하기 위해서 입니다. 
질문주신 의도는 유저가 a,b 라는 형태로 응답을 하지 않을 때도 있지 않을까 하는 말씀 이실 것 같습니다. 

유저의 응답을 받을 때  "게임 시작을 위해 최소 값, 최대 값을 입력해주세요. (예: 1, 50) "와 같이 
예시가 작성 되어있으므로 정상 작동을 위해서는 콤마가 들어간 형태로 작성이 되어야 한 다는 것을 유저가 인지 할 것 같지만 
다른 특수문자가 들어간다면 에러 날 여지가 있을 것 같습니다. 

그러나 모든 유저의 응답은 validation를 거치고 있으므로 
특수문자가 들어간다면 해당 오류날 여지를 거르고 있으므로
엣지케이스 대응이 되어 있는 걸로 판단됩니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+    const inputValid = validation(inputValue);
+    if (!inputValid) continue;
+
+    if (parseFloat(inputValue) === answer) {
```
ask;

parseFloat로 파싱한 이유가 있을까요? parseInt나 Number도 있을 것 같아서요.

codingjamee replied:
https://github.com/codingjamee/js-number-up-and-down/pull/2#discussion_r1623335989
이부분에 ask하신 부분에서 답변드린 이유로 parseFloat으로 사용 해보았습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+
+    if (parseFloat(inputValue) === answer) {
+      printResult({ result: "success", trial });
+      askRestart(); //여기다 둬도 동작 할지 확인
```
p5;

askRestart 함수가 비동기 함수인데 await를 넣지 않아도 괜찮을까요?

codingjamee replied:
askRestart함수가 비동기이지만, 해당함수를 호출하면 실행컨텍스트가 생성되고 
해당 함수의 모든 로직이 완성된 뒤에 break를 할 필요는 없을 것 같습니다. 
그 이유는 결괏값을 받아 그 값에 따라 로직이 변경되는 것이 아니기 때문입니다. 
askRestart함수를 호출하는 것으로 해당 if블록의 역할은 끝났고 더이상 지속되어야 할 로직도 없다고 판단됩니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+  return true;
+}
+
+async function tryLoop(trial, trialLimit, answer) {
```
P4;

해당 코드는 로직을 명확하게 분리 해볼 여지가 있을 것 같아요! 
요 함수를 조금 더 읽기 쉽게 바꿔 볼 수 있을까요?

codingjamee replied:
좀더 고민해보겠습니다.

codingjamee replied:
> P4;
> 
> 해당 코드는 로직을 명확하게 분리 해볼 여지가 있을 것 같아요! 요 함수를 조금 더 읽기 쉽게 바꿔 볼 수 있을까요?  

좀더 고민해보겠습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,138 @@
+}
+
+function validation(...arg) {
+  for (let i = 0; i < arg.length; i++) {
```
P3;

개인적으로 arr의 경우 고차배열 함수가 코드의 가독성이 훨씬 높아진다고 생각하기 때문에 선호하는 편이에요. 
고차 배열 함수로 바꿔보는 것은 어떨까요?

codingjamee replied:
> P3;
> 
> 개인적으로 arr의 경우 고차배열 함수가 코드의 가독성이 훨씬 높아진다고 생각하기 때문에 선호하는 편이에요. 고차 배열 함수로 바꿔보는 것은 어떨까요?

네 고차배열함수로 변경 해보았습니다.

eastroots92 reviewed (commented):
제인님 고생 많으셨어요!

지난주에 비해 의도가 훨씬 잘 읽혀서 좋았어요!! :+1:

문제에 대한 접근 방식은 매우 훌륭하다고 생각되나 풀어낸 방식에 있어서 아쉬운 부분이 있어 코멘트 남겨 두었어요.

크게 두가지인데.
1. 도메인 로직과 UI로직에 대한 분리 (Domain 쪽 코드에 코멘트를 남겨 두었어요.)
2. 함수의 역할을 명확하게 하고 의도가 예측 가능하게 하기.(또는 의도에 맞게 만들어보기)

혹시 해보시고 궁금한 점이 있다면 편하게 물어봐주세요!!

codingjamee committed:
Refactor: userInputSet함수 return값 나누어 사용해보기

codingjamee committed:
Refactor: printConsole함수 작성 및 유효하지않은 입력 playGame에서 다루기

codingjamee committed:
Refactor: validation함수 고차배열함수로 바꾸기

codingjamee committed:
console UI부분 로직 분리

codingjamee committed:
업다운 console로직 분리

codingjamee committed:
Refactor: 1. 함수 결합 2. loop함수 분리
1. 비슷한 로직인 두 개의 함수 결합
2. loop함수 명확히 하기 위해 playGame에서 결과 return, askRestart 호출

codingjamee committed:
askRestart 함수에서 printConsole함수를 인수 설정하여 호출

codingjamee committed:
Refactor: consoleContent함수로 result 설정

codingjamee committed:
Refactor: tryLoop 조건문 변수로 설정 및 한 줄로 설정

codingjamee committed:
종료문구 consoleContent함수에서 받아오기

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+    if (overLimit) ({ result: "fail" });
+
+    const guesses = guess.concat(gameData().showInput(guess, inputValue));
+    printConsole({ result: `이전 추측: ${guesses.join(", ")}` });
```
P4;

네이밍 규칙에 있어서 예측가능하게 네이밍을 하기 위한 통상적인 규칙들이 있어요. 예를들면 
1. 함수는 동사 형태로 작성한다.
2. 변수명은 명사형태를 유지한다. 
등등 이요. [(참고)](https://sooftware.io/coding-habit/)

이 이외에도 함수와 변수의 이름이 의도를 명확히 드러내는 것이 좋아요.
그렇기 떄문에 printConsole 대신 displayMessage 와 같은 네이밍이 더 좋지 않을까 생각했어요.
consoleContent 보다는 getConsoleContent 같은 네이밍이 더 좋을 것 같구요.

해당 부분에 대해서 어떻게 생각하시나요?

codingjamee replied:
네이밍규칙 자료 공유 감사합니다. 잘 읽어보고 반영하겠습니다. 

getConsoleContent는 함수의 이름 앞에 동사를 넣어야 하기 때문으로 이해되었습니다. 
그런데 printConsole대신 displayMessage가 더 낫다고 말씀주신부분의 의도가 잘 파악되지 않습니다. 

- print와 display의 차이가 어떤 것인지 조금더 구체적으로 말씀해주시면 감사하겠습니다.
 print는 출력한다는 의미이고 display는 나타낸다는 의미인데, 
 추후 리팩토링을 위해 현재는 console이지만 
 포괄적으로 사용할수 있으면서 의도가 명확하므로
이 부분이 낫다는 의도이실까요? 
제가 사용한 용어는 console에 국한되는 지엽적인 용어이기 때문일까요? 
- console은 개발용어이기 때문일까요? 
혹시 개발용어가 의도가 잘 드러나지 않는다는 의도이시라면
getConsoleContent의 네이밍도 console이란 개발 용어를 수정하는 것이 더 나을까요?

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+}
+
+async function startSetting() {
+  const { result: boundary, isValid: boundaryIsValid } = await userInputSet(
```
P4;

위에 언급했던 네이밍에 있어서 개인적으로는 userInputSet 보단 promptUser가, setMinMax보다는 parseBoundary가 좋지 않을까 생각해요!

codingjamee replied:
네! 감사합니다.
그런데 혹시 promptUser와 같은 선상에서 생각해보면 
user에게 game Infomation을 받는 것이니까 
promptGameInfo와 같은 네이밍은 어떨까요?
어떤 역할을 하는 함수인지 더 명확히 드러내보고자 위와 같이 지어보았습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+  return [minNumber, maxNumber];
+}
+
+async function userInputSet(askConstant) {
```
P3;

현재는 userInputSet이 너무 특정 값이 들어올 것을 가정하고 있는 것 같아요.
예를 들면 `,`로 split을 한다던지, 내부에서 숫자에 대한 `validation`을 한다던지요!

이런 경우 두가지 측면에서 고민이 드는 것 같아요.

1. 네이밍이 애매모호하다.
만약 함수의 내용대로 동작하게 할 것이었다면 네이밍에서 숫자를 입력 받고 return 하게 하는 함수라는 내용이 유추가 가능해야 할 것 같아요. 
의도는 이해가 되었으나 result와 isValid를 ruturn 하게 되어 result를 쓸 때 항상 isValid를 확인해야 하는 상황일까? 하는 고민돌 살짝 드네요..!

2. 함수가 여러가지 역할을 하고 있다.
만약 네이밍대로 함수의 동작을 유추한다면
단일 책임 원칙을 지키고 있을까? 하면 아닌 것 같아요.
현재 함수는 숫자에 대한 유효성 검사, 입력값을 받는 것 두가지 역할을 하고 있다고 생각해요.

```
const trialLimit = await promptUser(constant.askTrialLimit);
  if (!isValidTrialLimit(trialLimit)) {
    ~~~
  }
  ```
  
  이런 식으로 풀어보는게 좋지 않았을까? 하는 생각이 살짝 들었어요.

codingjamee replied:
사실 promptUser 굳이 함수로 빼낼 필요가 있을까?
하는 부분을 고민하다가 빼냈습니다. 
그런데 그렇게 빼내다보니 validation을 함께 빼내서 코드의 길이를 줄이는데만 
집중을 하여 단일책임 원칙에 위배되는 함수가 되었네요!

사실 validation을 startSetting부분에서 if문의 인라인으로 넣는다고 가정한다면,
굳이 promptUser라는 함수가 존재할 필요가 없을 것 같다는 생각이 드는데, 
멘토님의 의견은 어떠신지 알고싶습니다!
readLineAsync는 그 자체로는 의도를 잘 파악할 수 없으므로
따로 함수 네이밍을 통해 의도를 파악하도록 하는 것이 더 나을까요? 

만일 함수 네이밍자체가 의도 파악하기 위한 것만이라면, 
차라리 readLineAsync라는 함수 자체를 promptUser라고 하는 것이 낫지 않을까요?


```javascript
async function startSetting() {
  const trialLimit = await promptUser(
    constant.askTrialLimit
  );
   if (!isValidTrialLimit(trialLimit)) {
    ~~~
  }
```

```javascript
function promptUser(query) { //readLineAsync에서 네이밍 변경
  return new Promise((resolve, reject) => {
    if (arguments.length !== 1) {
      reject(new Error("arguments must be 1"));
    }

    if (typeof query !== "string") {
      reject(new Error("query must be string"));
    }

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    rl.question(query, (input) => {
      rl.close();
      resolve(input);
    });
  });
}
```

이런식으로요.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+  console.log(result);
+}
+
+async function startSetting() {
```
위에 `userInputSet`에 언급 한 것 처럼 startSetting 함수 또한 함수 및 변수의 역할 분리가 애매모호하다고 생각해요.

- 단일 책임 원칙
- 유효성 검사 분리를 하게 된다면 더욱 명확한 함수가 되지 않을까 싶어요.

아래는 임의로 만들어본 예제 코드에요.
```
async function startSetting() {
  const boundary = await promptUser(constant.askBoundary);
  const [minNumber, maxNumber] = parseBoundary(boundary);
  if (!isValidBoundary(minNumber, maxNumber)) {
    return { settingIsNotValid: true };
  }
  
  const trialLimit = await promptUser(constant.askTrialLimit);
  if (!isValidTrialLimit(trialLimit)) {
    return { settingIsNotValid: true };
  }
  
  const answer = gameData(minNumber, maxNumber).answer;
  return {
    answer,
    trialLimit,
    minNumber,
    maxNumber,
    settingIsNotValid: false,
  };
}

```

해당 코드를 무조건 사용하기 보다는 제인님이 생각하셨을 때 어떤 지 쓰윽 생각해보시고 이야기 나눠보고 싶어요.

codingjamee replied:
```javascript
if (!isValidBoundary(minNumber, maxNumber)) {
    return { settingIsNotValid: true };
  }
```

이부분에서 객체를 return하신 의도는 어떤 건지 알 수 있을까요? 잘 파악되지 않습니다.
혹시 
```javascript
async function startSetting() {
  const boundary = await promptUser(constant.askBoundary);
  const [minNumber, maxNumber] = parseBoundary(boundary);
  let settingIsNotValid;
  if (!isValidBoundary(minNumber, maxNumber)) {
    return settingIsNotValid=true
  }
```
이런 의도이실까요? 
아니면 제가 아직 이해하지 못한 부분이 있을까요? 

주신 코드를 스코프의 관점으로 파악해보면
함수 내부의 if블록내에서 settingIsNotValid라는 키값을 가지고 있는 객체를  return하게 되면 
객체 의 키 값 settingIsNotValid 변수는 startSetting에서 생성된 객체 내부의 스코프 이므로 
startSetting라는 상위 스코프에서는 접근하지 못할 것 같습니다.

codingjamee replied:
그리고 단일 책임의 원칙에 따르면
startSetting 함수도 똑같이 settingIsNotValid라는 두가지 동작이 되는 것인데, 
이부분은 괜찮은 걸까요? 
playGame도 여러가지 printConsole(수정 해주신 displayMessage)이라든지
여러가지 역할을 하고 있는데 
결국 함수는 한개의 책임만을 할 수 없는 것은 아닌가요? 
그렇다면 단일 책임의 원칙을 따르는 함수와 그렇지 않아도 되는 함수의 구분은 어떤 것인가요?

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+  return isValid;
+}
+
+async function tryLoop(trialLimit, answer) {
```
P5;

함수를 만들때 인자를 최소화 하라고 권장하고 있어요.
파라메터를 받지 않거나 하나로 받는 것, 더 파라메터가 많아지는 경우 인자를 객체화 하라는 이야기가 있는데요!

왜 이런 이야기가 나왔는지 쓰윽 고민해보셔도 좋을 것 같아요.
(인자를 최소화 하라는 것이 대체로 좋다이지 무조건 하나, 받지 않게 하라는 뜻은 아니에요!)

codingjamee replied:
리팩토링 책에서도 동일한 내용을 읽어서 고민이 되긴 했었습니다. 
인자를 여러개를 받으면 함수간의 의존성이 분리가 되지 않아서
리팩토링 할때나 테스트할때도 좋지 않다고 들었습니다. 
저는 그래서 결국 파일 모듈 스코프 내에서 prompt로 받는 유저의 데이터를
모든 함수들이 재사용 할 수 있도록 작성해야 할 것 같다는 생각이 들었습니다. 

한번 고민해서 반영해보겠습니다.

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+  const guess = [];
+  while (trial <= trialLimit) {
+    const inputValue = await readLineAsync("숫자 입력: ");
+    const inputValid = validation(inputValue);
+    if (!inputValid) continue;
+    const inputCorrect = parseFloat(inputValue) === answer;
```
P5;

공백 라인을 활용하여 논리적 단위를 분리하면 가독성이 높아져요. 
예를 들어, 함수 내부에서 주요 논리 블록 간에 공백을 추가하면 좋아요.

codingjamee replied:
넵 공백 라인 활용 해보겠습니다.!

eastroots92 reviewed (approved):
제인님 고생 많으셨어요.
해당 PR에서 이야기 하고 싶은 부분에 대해 코멘트를 남겨 두었어요.


일부 코드에 함수의 역할, 네이밍 이야기를 하였지만 전반적으로 고민해보시면 좋을 내용이라고 생각해요.

이후 PR에서도 이어서 리뷰가 진행 될 예정이기 때문에 추가로 PR리뷰 진행이 아닌 지금 제가 단 코멘트를 고민해보시고 적용한 다음에 스윽 머지해주시면 좋을 것 같아요.
(다만 제가 질문을 남긴 부분에 있어서는 꼭 코멘트를 남겨주세요.)

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,158 @@
+  const minNumber = Math.min(first, second);
+  const maxNumber = Math.max(first, second);
+  return [minNumber, maxNumber];
+}
```
1. 리드라인 어싱크 => 값을 입력 받는 친구. (아무 맥락 없음  그냥 입력 받아서 값을 전달하는 역할만 함.)

2. 제인님의 userInputSet
  1. 질문을 던진다.
  2. 질문에 대한 입력 값을 받는다.
  
---
  3. validation을 한다.
  
  
handleInputSet("질문", validation 함수 (return은 무조건 boolean이어야 함), {onSuccess, onError}) => {
  // 질문 출력
  값 입력 받고,
  if (validation(값)) {
  	return onSuccess(값);
  };
  
  onError();
});

handleInputSet('숫자 1~50 사이 입력해주세요.', 숫자인가 벨리데이션하는 함수, 성공시 로직, 실패시 로직);

codingjamee committed:
함수 네이밍 변경
consoleContent -> getGameStatusMessages로
printConsole -> displayMessage로
함수의 역할 명확히 하기 위해

codingjamee committed:
Refactor : 의도를 명확하게 하는 함수설정 및 책임소재 구분 해보기

codingjamee committed:
Refactor: 함수명 변경validation -> numberValidation

codingjamee committed:
Refactor: 함수명 변경 및 변경된 함수 사용
1. tryLoop -> executeGuessingGame : 알아맞추기 게임을 실행하므로
2. promptValidateGameInfo -> promptValidateUserInput :
gameInfo뿐만 아니라 재시작시 yes/no도 받으므로
3. promptValidateGameInfo의 인수 askConstant -> askContent :
질문 내용이므로
4. promptUser함수 추가:
질문내용을 받아서 받은 응답값을 return하는 함수
5. 139line: readLineAsync와 valid확인 함수 등을 promptValidateUserInput함수로 사용
6. isAnswerCorrect 생성 (167 line)

codingjamee committed:
Fix: return 빠진부분 추가

codingjamee commented:
### 시도한 바
- 함수 명 의도가 잘 드러나도록 수정하였습니다. 
- promptValidateUserInput의 책임을 사용자측에서 할 수 있도록 
   따로 인수로 받아서 처리할 수 있게 추상화 하였습니다. 
   (ex: askContent, validateFn, onInValid)
- 도메인 로직이라 판단되는 numberValidation 등을 domain으로 분류하였습니다. 

### 고민 되는 점
- askAndRestart부분은 메시지를 출력하고 응답을 받은 뒤 
결과에 따라 함수를 호출하는 함수인데, 이부분은 좀더 추상화를 하게 된다면 
조금더 보편적으로 쓸 수 있지 않을까 하는 생각이 듭니다. 
- boundary나 trialLimit이 유효하지 않을 경우 displayMessage와 askAndRestart를 하고 있는데, 
아예 게임을 끝낼 수 있는 방법은 무엇이 있을까 고민 중입니다. 
- executeGuessingGame을 좀더 추상화 해서 잘 읽히도록 하는 방법이 없을까 고민이 됩니다.

codingjamee committed:
numberValidation함수 domain 파일로 분류

codingjamee committed:
reduce->every 콜백함수로 변경

codingjamee committed:
Refactor: askRestart함수 인수 설정

codingjamee committed:
Refactor: messageFn -> restartMessageFn 이름 변경

codingjamee committed:
Refactor: askRestart -> askAndRestart 함수명 변경

codingjamee committed:
Fix: 호출시 parseFloat으로

codingjamee committed:
Refactor: parseBoundary함수 domain으로 분리

codingjamee committed:
UI에서 parseBoundary함수 삭제

codingjamee committed:
refactor: isAnswerCorrect를 domain으로 분리

eastroots92 commented on src/UI/index.js:
```diff
@@ -0,0 +1,188 @@
+  restart: "게임을 다시 시작합니다.",
+};
+
+const getGameStatusMessages = ({
```
P3;
문득 코드를 보다 해당 코드는 휴먼에러가 나기 너무 좋은 구조라는 생각이 들었어요.

예를들어 `getGameStatusMessages({ notValidType }).guessNumber` 를 넣으면 의도와 다른 동작이 발생하게 되니까요.
특히 JS이다보니 더욱이 메시지를 호출 할 때 어떤 값을 넘겨줘야하는지 매번 코드를 자세히 봐야 알 수 있어서 더 비효율이라 생각했던 것 같아요. `getGameStatusMessages.guessNumber({ minNumber, maxNumber });` 요런식으로 사용하는 것은 어떤가요?
